doc: net: Add OpenThread integration description (#2334)
nrfconnect/sdk-nrf: doc/nrf/ug_thread_ot_integration.rst

.. _openthread_integration:

OpenThread Integration
######################

.. contents::
    :local:
    :depth: 2

> The goal of this page is to explain how the OpenThread stack is integrated with remaining parts of the system in terms of the file structure,
> used threads and synchronization primitives and the network traffic flow.

Reviewer: ```suggestion
This page explains how the OpenThread stack is integrated with Zephyr and the nRF Connect SDK. The location of components in the ffile structure, used threads and synchronization primitives, and the network traffic flow are described.
```

Author: Will commit this change manually (`ffile` ;) )

> The OpenThread stack is integrated with the Zephyr RTOS as the L2 layer with the IEEE 802.15.4 Radio Driver underneath and Zephyr IPv6 stack above it.

Reviewer: Does it help with an L2/L3 + Zephyr IPv6/UDP/TCP diagram or is this obvious?

Author: Well, If I was looking at this page for the first time I would find this kind of description useful before starting the analysis of diagrams. Would you like to get rid of this part?

Reviewer: No - I was suggesting adding a diagram of L2, L3 (I assume there is an L1) if that makes sense.

Author: For the end user/application developer is it not needed IMO. If the Radio Driver Team decides to add more documentation they could potentially write a few words about L1. We mostly work with L2/L3/L4. The first layer is a bit outside of my scope.

> The OpenThread stack is integrated with the Zephyr RTOS as the L2 layer with the IEEE 802.15.4 Radio Driver underneath and Zephyr IPv6 stack above it.

Reviewer: ```suggestion
The OpenThread stack is integrated with the Zephyr RTOS as the L2 layer. The L2 layer uses the IEEE 802.15.4 Radio Driver.  The OpenThread API and IPv6 stack can be used directly, but by default, the Zephyr IPv6 stack is used.
```

> The main advantage of this approach is that it can utilize the Zephyr's L3 layer which allows for easy easy application porting and utilization of UDP, TCP and ICMP build on top of Zephyr's IPv6 stack.

Reviewer: ```suggestion
The advantage of using the Zephyr IPv6 stack is that it can utilize the Zephyr's L3 layer which allows for application portability and utilization of UDP, TCP and ICMP built on top of Zephyr's IPv6 stack.
```

> The drawback, on the other hand, lies in the reception path in which the IP packet needs to traverse the OpenThread L3 layer and then Zephyr's L3 layer in order to reach the BSD socket.

Reviewer: ```suggestion
The disadvantage is that the reception path where IP packets need to traverse the OpenThread L3 layer and then Zephyr's L3 layer in order to reach the BSD socket.
```

> .. note:: If one desires the application can use just the OpenThread API with it's IPv6 stack and even the internal CoAP implementation which is the application protocol.

Reviewer: ```suggestion
.. note:: The application can use the OpenThread API and it's IPv6 stack directly, including the internal CoAP implementation which is the application protocol.
```

Author: This part could be more precise.
Using OpenThread's IPv6 is not that easy. However using the UDPv6 based on that is.

I will just commit this version with `and it's UDPv6/IPv6 stack directly`.
If you think it should be rephrased please let me know.

> .. note:: If one desires the application can use just the OpenThread API with it's IPv6 stack and even the internal CoAP implementation which is the application protocol.

Reviewer: Seems to be an issue with this note where it runs into the paragraph below when rendered.  Is that interntional?

Author: It is not. 
@greg-fer, do you know how to fix that?

Reviewer 2: Notes should be added like this:
```suggestion
.. note::
   If one desires the application can use just the OpenThread API with it's IPv6 stack and even the internal CoAP implementation which is the application protocol.
```

>           Such approach reduces the overhead, becouse the packet does not need to go through the Zephyr's stack. 

Reviewer: ```suggestion
          This approach reduces the overhead, because the packet does not need to go through the Zephyr's stack in addition. 
```

>           This could lead to createing a quicker or more energy efficient application.

Reviewer: ```suggestion
          This result in a more energy efficient application with faster packet handling.
```

>           However it comes at at a cost of limited portability and functionality.

Reviewer: ```suggestion
          However it comes at the cost of limited portability and functionality.
```

>           Higher level protocols provided by OpenThread are not as advanced comparing to Zephyr's.

Reviewer: ```suggestion
          Higher level protocols provided by OpenThread are not as full-featured compared to those implemented in Zephyr.
```

          For example, CoAP implementation does not support observers and the block transfer functionality.

>           Moreover the TCP protocol is not available at all.

Reviewer: ```suggestion
          The TCP protocol is not is not implemented in OpenThread.
```

File system and shim layer
**************************

> The OpenThread Network Stack is present in the following paths:

Reviewer: ```suggestion
The OpenThread Network Stack components are located in the following paths:
```

* OpenThread stack location: ``modules/lib/openthread/``
* OpenThread shim layer: 
  
  * Thread entry point function, callbacks, utils, L2 registration: ``zephyr/subsys/net/l2/openthread/``
  * OpenThread platform layer location: ``zephyr/subsys/net/lib/openthread/platform/``

The shim layer responsibilities include:

* translating the data into the Zephyr's native net_pkt,
* providing openthread thread body and synchronization API,
* providinng ``openthread_send()``, ``openthread_recv()`` calls which are registered as the L2 interface API,
* providing a way to initialize the OpenThread stack,
* implementing callback functions used by the OpenThread stack

> The nRF IEEE802.15.4 Radio Driver is present in the following paths:

Reviewer: ```suggestion
The nRF IEEE802.15.4 Radio Driver is located in the following paths:
```

* nRF IEEE802.15.4 Radio Driver shim layer location: ``zephyr/drivers/ieee802154/{ieee802154_nrf5.c/ieee802154_nrf5.h}``
* nRF IEEE802.15.4 Radio Driver location: ``modules/hal/nordic/drivers/nrf_radio_802154``
The RX connection of the radio driver is done with the interrupt handler registered using the Zephyr's mechanism with the ``NRF_802154_INTERNAL_RADIO_IRQ_HANDLING=0`` defined. 
The registered IRQ handler uses Zephyr's FIFO to pass the IEEE802.15.4 frame further.
The ``802154 RX`` thread runs on the highest cooperative priority and waits on this FIFO. 
Once a new frame appears, it continues with the processing.

The TX connection uses the workqueue, which calls the radio driver calls to schedule the transmission.
Then the RTC IRQ is used to send the frame over the air.

Threads
*******

* ``openthread`` - responsible for receiving IEEE 802.15.4 frames during reception.

>   When the reassembled traffic turns out to be the IPv6 packet, it calls the ``ot_receive_handler()`` which injects it back to the L2 via the ``net_recv_data()``, so that it could later reach the Zephyr's IP stack.

Reviewer: ```suggestion
  When the reassembled traffic is an IPv6 packet, it calls the ``ot_receive_handler()`` which injects it back to the L2 via the ``net_recv_data()``, so that it could later reach the Zephyr's IP stack.
```

* ``workqueue`` - responsible for invoking the radio driver API in order to schedule a transmission.

* ``802154 RX`` - resposible for the "upper half" processing of the radio frame reception. 
  Works on the objects of type ``nrf5_802154_rx_frame`` which are put to the ``nrf5_data.rx_fifo`` from the RX IRQ context. 
  The thread is responsible of creating the net_pkt structure and passing it the upper layer with the ``net_recv_data()``.

Traffic flow
************
The traffic flow is not fully symmetrical for the reception (RX) and the transmission (TX) cases.
Each of these flows is described below in an appropriate section.

RX path
=======
An application typically consists of one or more :ref:`threads <threads_v2>`
that execute the application logic.
The following figure shows what happens when you are using the :ref:`BSD socket API <bsd_sockets_interface>`.

.. figure:: zephyr_netstack_openthread-rx_sequence.svg
    :alt: OpenThread Application RX data flow
    :figclass: align-center

    OpenThread Application RX data flow

Data receiving (RX) processing flow:

> 1.  A network data packet is received by a nRF5 IEEE 802.15.4 Radio Driver.

Reviewer: ```suggestion
1.  A network data packet is received by the IEEE 802.15.4 Radio Driver.
```

The stack is integrated with the Zephyr RTOS as the L2 layer with the IEEE 802.15.4 Radio Driver underneath and Zephyr IPv6 stack above it.
The advantage of using the Zephyr IPv6 stack is that it can utilize the Zephyr's L3 layer which allows for application portability and utilization of UDP, TCP and ICMP built on top of Zephyr's IPv6 stack.
The disadvantage is that the reception path where IP packets need to traverse the OpenThread L3 layer and then Zephyr's L3 layer in order to reach the BSD socket.

.. note:: The application can use the OpenThread API and it's IPv6 stack directly, including the internal CoAP implementation which is the application protocol.
          This approach reduces the overhead, because the packet does not need to go through the Zephyr's stack in addition.
          This result in a more energy efficient application with faster packet handling.
          However it comes at at the cost of limited portability and functionality.
          Higher level protocols provided by OpenThread are not as full-featured compared to those implemented in Zephyr.
          For example, CoAP implementation does not support observers and the block transfer functionality.

>           The TCP protocol is not is not implemented in OpenThread.

Reviewer 3: ```suggestion
          The TCP protocol is not implemented in OpenThread.
```

Author: Done, thanks!

Apart from these threads, the OpenThread stack also uses the :ref:`application threads <threads_v2>`.
Usually one or more, these threads execute the application logic.

File system and shim layer
**************************

The OpenThread network stack components are located in the following directories:

* OpenThread stack: :file:`modules/lib/openthread/`
* OpenThread shim layer:

  * Thread entry point function, callbacks, utils, L2 registration: :file:`zephyr/subsys/net/l2/openthread/`
  * OpenThread platform layer location: :file:`zephyr/subsys/net/lib/openthread/platform/`

The responsibilities of the OpenThread shim layer are as follows:

* Translating the data into the Zephyr's native :c:type:`net_pkt` structure.
* Providing the OpenThread thread body and synchronization API.
* Providing :cpp:func:`openthread_send()` and :cpp:func:`openthread_recv()` calls that are registered as the L2 interface API.
* Providing a way to initialize the OpenThread stack.
* Implementing callback functions used by the OpenThread stack.

The nRF IEEE 802.15.4 radio driver is located in the following directories:

> * nRF IEEE 802.15.4 radio driver shim layer: :file:`zephyr/drivers/ieee802154/{ieee802154_nrf5.c/ieee802154_nrf5.h}`

Author: ```suggestion
* nRF IEEE 802.15.4 radio driver shim layer: :file:`zephyr/drivers/ieee802154/{ieee802154_nrf5.c,ieee802154_nrf5.h}`
```
